### PR TITLE
[FLINK-22961][streaming] firstBarrierArrivalTime is recalculated when…

### PR DESCRIPTION
**Pull request by akalash**
… the first barrier or the barrier announcement received

## What is the purpose of the change

*PR fixes the moment since when alignment timeout should start *

## Brief change log

  - *firstBarrierArrivalTime recalculated when the checkpoint barrier or the barrier announcement was received rather than as in the past when only the barrier announcement was considered*

## Verifying this change

This change added tests and can be verified as follows:

  - *Added new test in AlternatingCheckpointsTest which takes into account the gate with Remote and Local channel*

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / **no**)
  - If yes, how is the feature documented? (not applicable / docs / JavaDocs / not documented)

### CODE REVIEW

**Comment by akalash** [comment]
@flinkbot run azure

**Comment by akalash** [comment]
@flinkbot run azure

File: flink-tests/src/test/java/org/apache/flink/test/checkpointing/IgnoreInFlightDataITCase.java [hunk 1]
```diff
@@ -117,12 +118,13 @@ private boolean executeIgnoreInFlightDataDuringRecovery() {
...
         env.disableOperatorChaining();
         env.getCheckpointConfig().enableUnalignedCheckpoints();
+        env.getCheckpointConfig().setAlignmentTimeout(Duration.ZERO);
         env.getCheckpointConfig().setCheckpointingMode(CheckpointingMode.EXACTLY_ONCE);
         env.getCheckpointConfig().setCheckpointIdOfIgnoredInFlightData(1);
         env.setRestartStrategy(fixedDelayRestart(1, 0));
 
         env.addSource(new NumberSource(lastCheckpointValue))
-                .shuffle()
+                .rebalance()
```
Comment (dawidwys) [inline flink-tests/src/test/java/org/apache/flink/test/checkpointing/IgnoreInFlightDataITCase.java:127]: Why is this necessary?

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/checkpointing/SingleCheckpointBarrierHandler.java [hunk 2]
```diff
@@ -255,7 +255,6 @@ public void processBarrierAnnouncement(
             CheckpointBarrier announcedBarrier, int sequenceNumber, InputChannelInfo channelInfo)
             throws IOException {
         if (checkNewCheckpoint(announcedBarrier)) {
-            firstBarrierArrivalTime = getClock().relativeTimeNanos();
             if (alternating) {
```
Comment (dawidwys) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/checkpointing/SingleCheckpointBarrierHandler.java:258]: Shouldn't we register the Alignment timer on the first barrier as well If there were no announcements for that barrier? I think we could move this block into the `checkNewCheckpoint` method.

**Comment by dawidwys** [comment]
Thanks for the fix! Good catch! I had one question/suggestion.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/checkpointing/SingleCheckpointBarrierHandler.java (hunk 2 above)
Comment (akalash) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/checkpointing/SingleCheckpointBarrierHandler.java:258]: Good point. I believe it indeed makes sense to move it to `checkNewCheckpoint`

File: flink-tests/src/test/java/org/apache/flink/test/checkpointing/IgnoreInFlightDataITCase.java (hunk 1 above)
Comment (akalash) [reply flink-tests/src/test/java/org/apache/flink/test/checkpointing/IgnoreInFlightDataITCase.java:127]: This test should guarantee that all three subtasks will receive at least one record. Right now, the source emits fifty records which are enough to make this test working even with `shuffle` but strictly to say `shuffle` doesn't make sense here while `rebalance` does exactly what we need. So this change is not really necessary but it provides more correct behaviour for this test.

Comment (dawidwys) [reply flink-tests/src/test/java/org/apache/flink/test/checkpointing/IgnoreInFlightDataITCase.java:127]: Why can't we use the default `forward` partitioner then?

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/checkpointing/SingleCheckpointBarrierHandler.java (hunk 2 above)
Comment (akalash) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/checkpointing/SingleCheckpointBarrierHandler.java:258]: As I understand, this will be only partially working anyway: if we received the first barrier from localChannel without announcement and the timeout happened, then only barriers that have the announcements can be prioritized but if the other channel is local there will be no effect.
So how do you think it is problem or not if we can prioritize the barriers only for the remote channel(which has an announcement)?

**Comment by akalash** [comment]
@dawidwys I pushed the new commit then I can squash it with the first one when we found the final solution.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/checkpointing/SingleCheckpointBarrierHandler.java (hunk 2 above)
Comment (dawidwys) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/checkpointing/SingleCheckpointBarrierHandler.java:258]: I think that's a valid yet, unrelated issue. It's the same independent if we first receive an announcement or a barrier for a given checkpoint.

Comment (dawidwys) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/checkpointing/SingleCheckpointBarrierHandler.java:258]: cc @pnowojski

File: flink-tests/src/test/java/org/apache/flink/test/checkpointing/IgnoreInFlightDataITCase.java (hunk 1 above)
Comment (akalash) [reply flink-tests/src/test/java/org/apache/flink/test/checkpointing/IgnoreInFlightDataITCase.java:127]: Perhaps, I don't understand something but who guarantees that all subtasks will receive the records? As I can see in the implementation of ForwardPartitioner it always returns the first subpartition(`return 0;`)

Comment (dawidwys) [reply flink-tests/src/test/java/org/apache/flink/test/checkpointing/IgnoreInFlightDataITCase.java:127]: Does your source have different parallelism than the following map operator? I don't think so. Using `ForwardPartitioner` will result in subtask 0 of the source pushing to subtask 0 of the map operator and so on. If your only goal is to have each of the map operators receive data, I think that's enough, isn't it? 

Rebalance means that each source subtask will produce to every map subtask.

Comment (akalash) [reply flink-tests/src/test/java/org/apache/flink/test/checkpointing/IgnoreInFlightDataITCase.java:127]: > Does your source have lower parallelism than the following map operator? 
Actually, it does. I have only one source and three followed maps.I did it because it is easier to manage the execution when you have only one source.

Comment (dawidwys) [reply flink-tests/src/test/java/org/apache/flink/test/checkpointing/IgnoreInFlightDataITCase.java:127]: Are you sure? Where do you set the parallelism to `1` for the source? I can't seem to find it.

**Comment by dawidwys** [comment]
BTW, I think you actually forgot to push the commit ;)

File: flink-tests/src/test/java/org/apache/flink/test/checkpointing/IgnoreInFlightDataITCase.java (hunk 1 above)
Comment (akalash) [reply flink-tests/src/test/java/org/apache/flink/test/checkpointing/IgnoreInFlightDataITCase.java:127]: Yes, I am sure. It happens because I implemented SourceFunction which always has parallelism 1 rather than ParallelSourceFunction which has the configured parallelism.

Comment (dawidwys) [reply flink-tests/src/test/java/org/apache/flink/test/checkpointing/IgnoreInFlightDataITCase.java:127]: Right, I keep forgetting about the `ParallelSourceFuntion` :facepalm: It's such a design failure imo.

Comment (dawidwys) [reply flink-tests/src/test/java/org/apache/flink/test/checkpointing/IgnoreInFlightDataITCase.java:127]: Still, I believe you don't need to specify the partitioning explicitly. If you remove the `rebalance` whatsoever it will use the `RescalePartitioner` which behaves the same as `RebalancePartitioner` in this case, because you have just a single source task.

If you do believe a `rebalance` is better, add a comment why you need the `rebalance` there. It's not straightforward to figure out the role of it.

Comment (akalash) [reply flink-tests/src/test/java/org/apache/flink/test/checkpointing/IgnoreInFlightDataITCase.java]: Ok, I see it works pretty well even without `rebalance`

**Comment by akalash** [comment]
@dawidwys let's just hold this review for a while because according to this ticket https://issues.apache.org/jira/browse/FLINK-23041 which I just created, perhaps, we don't need `firstBarrierArrivalTime` at all.

File: flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/CheckpointOptions.java [hunk 3]
```diff
@@ -47,7 +47,11 @@
         FORCED_ALIGNED
     }
 
-    public static final long NO_ALIGNMENT_TIME_OUT = Long.MAX_VALUE;
+    /**
+     * Long.MAX_VALUE can not be used because this value should represent ms which will be converted
+     * to ns.
+     */
+    public static final long NO_ALIGNMENT_TIME_OUT = 2L * Integer.MAX_VALUE;
```
Comment (akalash) [inline flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/CheckpointOptions.java:54]: Oh, I actually think it isn't allowed to do so. I need to rethink it.

File: flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/CheckpointOptions.java [hunk 4]
```diff
@@ -47,7 +48,8 @@
         FORCED_ALIGNED
     }
 
-    public static final long NO_ALIGNMENT_TIME_OUT = Long.MAX_VALUE;
+    /** The maximum possible value is Long.MAX_VALUE 'ns' but API receives 'ms'. */
+    public static final long NO_ALIGNMENT_TIME_OUT = Duration.ofNanos(Long.MAX_VALUE).toMillis();
```
Comment (dawidwys) [inline flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/CheckpointOptions.java:52]: What is the problem with the old value? After all, it's just a marker value. It does not have any true meaning.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/checkpointing/SingleCheckpointBarrierHandler.java [hunk 5]
```diff
@@ -299,18 +294,23 @@ private void registerAlignmentTimer(CheckpointBarrier announcedBarrier) {
...
         long barrierId = barrier.getId();
-        if (currentCheckpointId < barrierId) {
-            if (isCheckpointPending()) {
-                cancelSubsumedCheckpoint(barrierId);
-            }
-            currentCheckpointId = barrierId;
-            numBarriersReceived = 0;
-            allBarriersReceivedFuture = new CompletableFuture<>();
-            return true;
+        if (currentCheckpointId >= barrierId) {
```
Comment (dawidwys) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/checkpointing/SingleCheckpointBarrierHandler.java:299]: Nice refactoring! I like returning early.

**Comment by dawidwys** [comment]
I have just a single question. Apart from that looks good.

File: flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/CheckpointOptions.java (hunk 4 above)
Comment (akalash) [reply flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/CheckpointOptions.java:52]: It was impossible to set NO_ALIGNMENT_TIME_OUT when alternating is enabled because when we are registering the register timer it converts this value to nanosecond which leads to the error because it was Long.MAX_VALUE.
We didn't see this problem because such a combination alternating + NO_ALIGNEMNT_TIME_OUT didn't meet before my changes but after I moved registerAlignmentTimer to checkNewCheckpoint some tests meet this problem.

**Comment by akalash** [comment]
@dawidwys. I squashed a couple of commits. And I answered your question.

File: flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/CheckpointOptions.java (hunk 4 above)
Comment (dawidwys) [reply flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/CheckpointOptions.java:52]: Hmm, ok I understand the problem, but then I think the solution is actually wrong. The `NO_ALIGNEMNT_TIME_OUT` means there should be no alignment timeout, therefore we should not register the timer.

We should rather add a condition in the `checkNewCheckpoint`:

```
   private void checkNewCheckpoint(CheckpointBarrier barrier) throws IOException {
        long barrierId = barrier.getId();
        if (currentCheckpointId < barrierId) {
            if (isCheckpointPending()) {
                cancelSubsumedCheckpoint(barrierId);
            }
            currentCheckpointId = barrierId;
            numBarriersReceived = 0;
            allBarriersReceivedFuture = new CompletableFuture<>();
            return true;
        if (currentCheckpointId >= barrierId) {
            return; // This barrier is not the first for this checkpoint.
        }

        if (isCheckpointPending()) {
            cancelSubsumedCheckpoint(barrierId);
        }
        currentCheckpointId = barrierId;
        numBarriersReceived = 0;
        allBarriersReceivedFuture = new CompletableFuture<>();
        firstBarrierArrivalTime = getClock().relativeTimeNanos();

        if (alternating && barrier.getCheckpointOptions().isTimeoutable()) {
            registerAlignmentTimer(barrier);
        }
        return false;
    }
```

Comment (akalash) [reply flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/CheckpointOptions.java]: good point. I made such a fix

**Review by dawidwys — APPROVED** [review]
LGTM
